### ISSUE TITLE
RFC - 6.0 - Improve Form template adjustability

### ISSUE THREAD

**Issue by LordSimal** [cakephp/cakephp#18088]
### Description

To adjust the generated HTML for `$this->Form->control()` calls one has to juggle between

* adjusting FormHelper template config for the used StringTemplates when loading in the AppView
* overwriting the FormHelper methods
* overwriting specific Widget classes

to get a desired HTML representation for a given `$this->Form->control()` (or other method) calls.

But this is still rather strict on what can be adjusted, since StringTemplate variables are a simple 1:1 replacement, not a Twig like template engine, since we just do a simple [vsprintf](https://github.com/cakephp/cakephp/blob/5.x/src/View/StringTemplate.php#L261)

Therefore you can't easily e.g. add a specific HTML attribute or class to an input/div, if a given templateVar is present or not. You always have to make it add that class or attribute, just with 2 different values (`is-visible` vs `is-hidden` CSS class or `data-is-visible` vs `data-is-hidden` as an attribute)

I know HTML attributes can easily be added per `$this->Form->control()` call, but what I am talking about is adjusting HTML depending on the given context and the type of input. 

---

The ultimate solution to this problem would be loading a simple PHP file per current string template with all the data (AND the context) passed to it, so that the user can decide for themselves what should be printed or not (via e.g. output buffering).

I am not sure what the performance impact would be here, but I need to finish my proof of concept first to get that benchmarked.

---

This kind of change would also make it easier for modern CSS libs like Tailwind to point into a directory and generate all the needed CSS classes since it (in very simple terms) looks for `class=".*"`) when generating the CSS. 
Currently some CSS classes live inside PHP class props/configs as well (see e.g. `errorClass` in the FormHelper) which can't easily be fetched by Tailwind Config without custom regex patterns.

To achieve this one would have to adjust all Helper and Widget classes to NOT contain any HTML related strings but just modify "state", which gets passed down to the template. The template in the end decides upon that state and context, what gets printed. Not the Helper, nor the Widget.

Of course this would come with a simple `bin/cake copy_form_templates` command to copy the default templates into your own app template and adjust accordingly.

With that I'd feel the whole Form Templating system would get a whole lot easier to adjust since its much clearer what is the "data loading/modification part" (Helpers + Widgets) and what is the actual printing part.

### CakePHP Version

6.0

**Comment by markstory** [comment]
> The ultimate solution to this problem would be loading a simple PHP file per current string template with all the data (AND the context) passed to it, so that the user can decide for themselves what should be printed or not (via e.g. output buffering).

Would this work similar to element rendering where we pass rendering off to a PHP template?

How do you see template files and the existing string templates interacting? If both a template file and template string are defined which one gets used?

**Comment by LordSimal** [comment]
> Would this work similar to element rendering where we pass rendering off to a PHP template?

yes. I wouldn't put those templates into the `elements` folder though to not get things mixed up. I see it more like a `config/form_templates`.

> How do you see template files and the existing string templates interacting? If both a template file and template string are defined which one gets used?

The already existing StringTemplate logic would be used as a fallback if a given widget template cannot be found.

**Comment by jamisonbryant** [comment]
Removed stale label due to recent activity/mention.

**Comment by jamisonbryant** [comment]
Could we move the existing string templates to PHP files (elements seems like a good fit) such as `input_container.php`? Then users could customize easily and also include other custom templates using additional elements. 

I think this ultimately might necessitate making the templates work a bit more like (mini-) views. Maybe get rid of string templates entirely and embrace widgets as the standard?

**Comment by ADmad** [comment]
Having one file per string template would create too much file clutter. For e.g. just `FormHelper` has over 30 templates and probably a similar amount or more combined for rest of the helpers.

**Comment by LordSimal** [comment]
Its not about moving the content purely into files, its more about the ability to overwrite the content for those templates via files (and therefore create a more dynamic way to control the HTML for the current state of the input)

**Comment by ADmad** [comment]
Okay, having a POC would be nice.

**Comment by LordSimal** [comment]
I have revisited this topic multiple times now but haven't found a way to make this as interoperable to the current StringTemplate system as I would like it to be (without major changes)

The main problem lies within the fact, that String templates are so "dumb" that they don't have context - which is necessary to display e.g. error messages for the FormHelper when errors are set for an entity form (which are the most common usecase I'd argue for our users). The FormHelper itself has the context, but not the StringTemplates. And passing it down optionally feels wrong since it clutters up a lot of places in the StringTemplate logic.

Therefore I will continue with my desired behavior in a separate plugin, make my own Helper without the StringTemplates and will close this issue for now.
